BIP 348: OP_CHECKSIGFROMSTACK (#1535)
bitcoin/bips: bip-csfs.md

| BIP-XXXX   | `CHECKSIGFROMSTACK`
| :--------- | :------------------
| Layer      | Consensus (soft fork)
| Author     | Brandon Black <freedom@reardencode.com>, Jeremy Rubin <j@rubin.io>
| Discussion | https://delvingbitcoin.org/t/lnhance-bips-and-implementation/376/9
| Status     | Draft
| Type       | Standards Track
| Created    | 2023-12-22

> | License    | BSD-3-CLAUSE

Reviewer: Please use the [preamble format](https://github.com/bitcoin/bips/blob/master/bip-0002.mediawiki#bip-header-preamble) specified in BIP2.

```
BIP: XXX
Layer: Consensus (soft fork)
Title: CHECKSIGFROMSTACK
Author: Brandon Black <freedom@reardencode.com>, Jeremy Rubin <j@rubin.io>
Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-XXXX
Status: Draft
Type: Standards Track
Created: 2023-12-22
License: BSD-3-CLAUSE
```

## Abstract

This BIP describes two new opcodes for the purpose of checking cryptographic
signatures in bitcoin scripts against data other than bitcoin transactions.

## Summary

We propose replacing `OP_NOP5` (0xb4) in bitcoin script with
`OP_CHECKSIGFROMSTACKVERIFY`. When verifying taproot script spends having
leaf version 0xc0 (as defined in [BIP 342]), we propose `OP_CHECKSIGFROMSTACK`
to replace `OP_SUCCESS204` (0xcc).

`OP_CHECKSIGFROMSTACK` and `OP_CHECKSIGFROMSTACKVERIFY`
have semantics similar to `OP_CHECKSIG` and
`OP_CHECKSIGVERIFY` respectively, as specified below.

Only 32-byte keys are constrained. Similar to [BIP 341] unknown key types, for
other key lengths no signature verification is performed.

## Specification

* If fewer than 3 elements are on the stack, the script MUST fail and terminate immediately.

> * The public key (top element), message (second to top element), and signature (third from top element) are read from the stack.

Reviewer 2: nit: I'd like to offer the suggestion that the push order goes:

`<msg> <sig> <key>` instead of `<sig> <msg> <key>`. The current `CHECKSIG` ordering is `<sig> <key>` and it seems odd to me that we would make the CSFS opcode put the msg *between* these other pushes. Alternatively you could go `<sig> <key> <msg>` and it would seem to me as equally intuitive. As is though it feels like a gotcha.

Is there a specific reason this choice was made that I hadn't considered?

Author: Anything that comes from the input stack should be before things that are provided by the locking script, so `sig` definitely comes first. Depending on particulars `msg` might come from the input stack or the locking script, and `key` always comes from the locking script. Examples:
* `<delegate-sig> <sig> <delegate-key> || DUP TOALT <key> CSFS FROMALT CHECKSIG`
* `<sig> <hash> || CTV <key> CSFS`
* `<sig> || <flags> TXHASH <key> CSFS`
* `<sig> <settlement-hash> <update-hash> || CTV SHA256 CAT <key> CSFS`

Reviewer 2: Retracted.

I understand the general direction of what you're getting at but I have to admit that I'm not following the specific use cases where `<msg>` might come from the witness instead of the script. It makes sense to me why you'd want `<msg>` in the script, what's lest clear is under what conditions it is useful to let the witness define it.

Author: In any kind of delegation (see the first example above) the message is provided by the witness (i.e. the thing you're delegating to by signing). Here's a powerful example:

`<delegate_sig> <sig> <delegate_pubkey> <locktime> || CLTV SHA256 CAT <pubkey> CSFS VERIFY CHECKSIG`

Delegate to a key but only after `<locktime>`, with different locktimes per delegate key.

Reviewer 2: Forgive me but I have to expand this out because it doesn't make sense to me:

Witness:
1. delegate signature is pushed onto the stack: `<delegate_sig>`
2. the signature authorizing the delegate is pushed: `<delegate_sig> <authorizing_sig>`
3. the delegate pubkey is pushed: `<delegate_sig> <authorizing_sig> <delegate_pubkey>`
4. locktime is pushed: `<delegate_sig> <authorizing_sig> <delegate_pubkey> <locktime>`

Script:
5. locktime is checked: `<delegate_sig> <authorizing_sig> <delegate_pubkey> <locktime>`
6. locktime is hashed: `<delegate_sig> <authorizing_sig> <delegate_pubkey> <locktime_hash>`
7. locktime hash combined with delegate pubkey: `<delegate_sig> <authorizing_sig> <delegate_pubkey || locktime_hash>`
8. authorizing pubkey is pushed: `<delegate_sig> <authorizing_sig> <delegate_pubkey || locktime_hash> <authorizing_pubkey>`
9. CSFS(V) operation(s) checks that the delegate pubkey is authorized after the locktime by checking the `key || locktime` message was signed by `authorizing_sig`: `<delegate_sig>`
10. CHECKSIG operation fails because we don't have a key? Maybe you forgot a dup somewhere? Maybe I don't understand script as well as I thought? idk.

Assuming this is roughly what you meant, then I'd imagine in this case the authorizing signature would be pre-signed and distributed. Then at transaction time the delegate presents the authorizing signature along with their key and assigned locktime.

Assuming all of this is a fair reading, then I think I get what you mean with `<msg>` having an interesting use case wherein it could be:
1. provided by the script
2. provided by the witness
3. collaboratively derived from witness and script participation.

... which seems pretty cool. I'd love if you can correct either my understanding or your example for completeness, but otherwise I'm satisfied with this and this thread can be resolved.

Author: Sorry for the long delay. You're quite correct that the example I gave does not work. Apparently I should have waited for Taproot Wizards new neovim script plugin. I believe this works for 3 more opcodes :(

`<sig> <delegate_sig> <delegate_pubkey> <locktime> || CLTV SHA256 OVER CAT TOALT CHECKSIGVERIFY FROMALT <pubkey> CSFS`

1. the signature authorizing the delegate is pushed: `<authorizing_sig>`
2. delegate signature is pushed onto the stack: `<authorizing_sig> <delegate_sig>`
3. the delegate pubkey is pushed: `<authorizing_sig> <delegate_sig> <delegate_pubkey>`
4. locktime is pushed: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime>`
5. locktime is checked: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime>`
6. locktime is hashed: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime_hash>`
7. delegate pubkey is copied over: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime_hash> <delegate_pubkey>`
8. locktime hash combined with delegate pubkey: `<authorizing_sig> <delegate_sig> <delegate_pubkey> <locktime_hash || delegate_pubkey>`
9. concatenation is moved to alt stack: `<authorizing_sig> <delegate_sig> <delegate_pubkey>`
10. CHECKSIGVERIFY validates the signature from the delegated key: `<authorizing_sig>`
11. concatenation is moved from alt stack: `<authorizing_sig> <locktime_hash || delegate_pubkey>`
12. authorizing pubkey is pushed: `<authorizing_sig> <locktime_hash || delegate_pubkey> <authorizing_pubkey>`
10. CSFS operation(s) checks that the delegate pubkey is authorized after the locktime by checking the `locktime || key` message was signed by `authorizing_sig`: `TRUE`

Thanks for checking the script.

```
BIP: XXX
Layer: Consensus (soft fork)
Title: CHECKSIGFROMSTACK
Author: Brandon Black <freedom@reardencode.com>, Jeremy Rubin <j@rubin.io>
Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-XXXX
Status: Draft
Type: Standards Track
Created: 2023-12-22
License: BSD-3-CLAUSE
```

## Abstract

This BIP describes two new opcodes for the purpose of checking cryptographic
signatures in bitcoin scripts against data other than bitcoin transactions.

## Summary

When verifying taproot script spends having leaf version 0xc0 (as defined in
[BIP 342]), we propose `OP_CHECKSIGFROMSTACK` to replace `OP_SUCCESS204`
(0xcc).

> `OP_CHECKSIGFROMSTACKVERIFY` has semantics similar to `OP_CHECKSIG`, as

Reviewer: This mentions OP_CSFSV, was that supposed to say just `OP_CHECKSIGFROMSTACK` or is this sentence obsolete?

specified below.

Only 32-byte keys are constrained. Similar to [BIP 341] unknown key types, for

> other key lengths no signature verification is performed.

Reviewer: This feels a bit out-of-context because keys have not been mentioned at all so far.

## Specification

* If fewer than 3 elements are on the stack, the script MUST fail and terminate immediately.
* The public key (top element), message (second to top element), and signature (third from top element) are read from the stack.
* The top three elements are popped from the stack.
* If the public key size is zero, the script MUST fail and terminate immediately.
* If the public key size is 32 bytes, it is considered to be a public key as described in [BIP 340]:
    * If the signature is not the empty vector, the signature is validated against the public key and message according to [BIP 340]. Validation failure in this case immediately terminates script execution with failure.

> * If the public key size is not zero, and it is not a [BIP 340] public key; the public key is of an unknown public key type, and no actual signature verification is applied. During script execution of signature opcodes they behave exactly as known public key types except that signature validation is considered to be successful.

Reviewer: I’m not sure whether this is intentional, but previously we treat a 32-byte key as a BIP 340 public key. I assume if it were out-of-range, the signature validation would fail.
However here, rather than repeating that the lengths 0-byte and 32-byte are excluded, the criterion appears to be "not zero and not a BIP 340 public key". I assume this is covered by the previous line, but what if a 32-byte value were provided that is out of range, would this line apply or not? 

Is it possible that this should read:

```suggestion
* If the public key size is not zero bytes and not 32 bytes; the public key is of an unknown public key type. Signature verification for unknown public key types succeeds as if signature verification for a known public key type had succeeded.
```

Author: I believe I copied that text from BIP340. Your wording is more clear.

* If the script did not fail and terminate before this step, regardless of the public key type:
    * If the signature is the empty vector: An empty vector is pushed onto the stack, and execution continues with the next opcode.
    * If the signature is not the empty vector:
        * The opcode is counted towards the sigops budget as described in [BIP 342].
        * A 1-byte value 0x01 is pushed onto the stack.

## Design Considerations

1. Message hashing: [BIP 340] is compatible with any size of message and does not require it to be a securely hashed input, so the message is not hashed prior to [BIP 340] verification.
2. Lack of verify semantics: Adding a single opcode for this purpose keeps the implementation and design simple. An earlier draft had a verify variant as a NOP upgrade, and if this funcationality is later brought to legacy scripts, that would be a good time to add a verify variant.

> 3. Add/multisig: No concession is made to `OP_CHECKMULTISIG` or `OP_CHECKSIGADD` semantics with `OP_CHECKSIGFROMSTACK(VERIFY)`. In Tapscript, add semantics can be implemented with 1 additional vByte per key (`OP_TOALTSTACK OP_CHECKSIGFROMSTACK OP_FROMALTSTACK OP_ADD`).

Reviewer: Is this "(VERIFY)" obsolete or shorthand for `OP_CSFS OP_VERIFY` here?

```suggestion
3. Add/multisig: No concession is made to `OP_CHECKMULTISIG` or `OP_CHECKSIGADD` semantics with `OP_CHECKSIGFROMSTACK`. In Tapscript, add semantics can be implemented with 1 additional vByte per key (`OP_TOALTSTACK OP_CHECKSIGFROMSTACK OP_FROMALTSTACK OP_ADD`).
```

> 4. Splitting R/S on the stack: Implementing split/separate signatures is left as an exercise for other bitcoin upgrades, such as `[BIP 347]`.

Reviewer: It would be easier to follow if the title of BIP 347 were mentioned:

```suggestion
4. Splitting R/S on the stack: Implementing split/separate signatures is left as an exercise for other bitcoin upgrades, such as [BIP 347] (OP_CAT).
```

It also isn’t clear to me why this link text is code formatted.

> 5. [BIP 118]-style Taproot internal key: Rather than introducing an additional key type in this change, we suggest implementing [BIP 349] or separately introducing that key type for all Tapscript signature checking operations in a separate change.

Reviewer: Ah, my bad, it’s not that it’s not merged, but that it is a markdown file rather than mediawiki.

> 5. [BIP 118]-style Taproot internal key: Rather than introducing an additional key type in this change, we suggest implementing [BIP 349] or separately introducing that key type for all Tapscript signature checking operations in a separate change.

Author: am confused, the PR shows merged.

> 5. [BIP 118]-style Taproot internal key: Rather than introducing an additional key type in this change, we suggest implementing [BIP 349] or separately introducing that key type for all Tapscript signature checking operations in a separate change.

Author: ah, had the link wrong - `md`

## Resource Limits

These opcodes are treated identically to other signature checking opcodes and
count against the various sigops limits and budgets in their respective script
types.

## Motivation

### LN Symmetry

When combined with [BIP 119] (`OP_CHECKTEMPLATEVERIFY`/CTV),

> `OP_CHECKSIGFROMSTACK` (CSFS) can be used in Lightning Symmetry channels.

Author: For the purposes of Lightning Symmetry, when not attempting to build an alternative Lightning Symmetry that has endogenous fees, CSFS+CTV are a minimal substitute for APO. CSFS+CTV+PC+IKEY (LNHANCE) are byte-wise and composability-wise better substitute to APO.

> `OP_CHECKSIGFROMSTACK` (CSFS) can be used in Lightning Symmetry channels.

Author: Added a brief summary of this to the BIP
